src/procedures/proc_bfs.c

```diff
@@ -0,0 +1,222 @@
+	if(reltype == NULL) {
+		R = Graph_GetAdjacencyMatrix(gc->g);
+		TR = Graph_GetTransposedAdjacencyMatrix(gc->g);
+	} else {
+		Schema *s = GraphContext_GetSchema(gc, reltype, SCHEMA_EDGE);
+		if(!s) return PROCEDURE_OK; // Failed to find schema, first step will return NULL.
+		pdata->reltype_id = s->id;
+		R = Graph_GetRelationMatrix(gc->g, s->id);
+		if(Config_MaintainTranspose()) TR = Graph_GetTransposedRelationMatrix(gc->g, s->id);
+		else TR = GrB_NULL;
+	}
+
+	GrB_Vector V;  // Vector of results
+	GrB_Vector PI = NULL; // Vector backtracking results to their parents.
+	GrB_Vector *PI_ptr = &PI;
+	if(!collect_edges) {
+		/* If we're not collecting edges, pass a NULL parent pointer so that the algorithm will
+		 * not perform unnecessary work. */
+		PI_ptr = GrB_NULL;
+		pdata->edges_output_idx = -1;
+	}
+	assert(LAGraph_bfs_both(&V, PI_ptr, R, TR, source_id, max_level, false) == GrB_SUCCESS);
+	/* Remove all values with a level less than or equal to 1.
+	 * Values of 0 are not connected to the source, and values of 1 are the source. */
+	GxB_Scalar thunk;
+	GxB_Scalar_new(&thunk, GrB_UINT64);
+	GxB_Scalar_setElement_UINT64(thunk, 1);
+	GxB_Vector_select(V, GrB_NULL, GrB_NULL, GxB_GT_THUNK, V, thunk, GrB_NULL);
+	GxB_Scalar_free(&thunk);
+
+	// Get number of entries.
+	GrB_Index nvals;
+	GrB_Vector_nvals(&nvals, V);
+	pdata->n = nvals;
+
+	// Retrieve all tuples representing connected nodes.
+	GrB_Index *node_ids = rm_malloc(nvals * sizeof(GrB_Index));
+	GrB_Vector_extractTuples_INT64(node_ids, GrB_NULL, &nvals, V);
+	pdata->node_ids = node_ids;
+	pdata->parents = PI;
+
+	GrB_Vector_free(&V);
+	return PROCEDURE_OK;
+}
+
+static SIValue *Proc_BFS_Step(ProcedureCtx *ctx) {
+	assert(ctx->privateData);
+
+	BFSContext *pdata = (BFSContext *)ctx->privateData;
+
+	// Return NULL if this source has already been mapped or there are no connected nodes.
+	if(pdata->depleted || pdata->n == 0) return NULL;
+
+	// Build arrays for the outputs the user has requested.
+	if(pdata->nodes_output_idx >= 0) pdata->output[pdata->nodes_output_idx] = SI_Array(pdata->n);
+	if(pdata->edges_output_idx >= 0) pdata->output[pdata->edges_output_idx] = SI_Array(pdata->n);
+
+	for(int i = 0; i < pdata->n; i ++) {
+		// Get the reached node.
+		NodeID node_id = pdata->node_ids[i];
+
+		if(pdata->nodes_output_idx >= 0) {
+			// Append each reachable node to the nodes output array.
+			Node n = GE_NEW_NODE();
+			Graph_GetNode(pdata->g, node_id, &n);
+			SIArray_Append(&pdata->output[pdata->nodes_output_idx], SI_Node(&n));
+		}
+
+		if(pdata->edges_output_idx >= 0) {
+			GrB_Index parent_id;
+			// Find the parent of the reached node.
+			assert(GrB_Vector_extractElement(&parent_id, pdata->parents, node_id) == GrB_SUCCESS);
+			parent_id --; // Decrement the parent ID by 1 to correct 1-indexing.
+			// Retrieve any edge connecting the parent node to the current node.
+			EdgeID id = Graph_GetSingleEdgeConnectingNodes(pdata->g, parent_id, node_id, pdata->reltype_id);
```

Comment:
To reduce code size and overloading graph.h/c with additional functionality I prefer we'll use `Graph_GetEdgesConnectingNodes` here.